Update/waku replace (#114)
status-im/specs: docs/stable/1-client.md

Protocol, v5](https://github.com/ethereum/devp2p/blob/master/rlpx.md) is used, which
allows for TCP-based communication between nodes.

On top of this we run the RLPx-based subprotocol [Waku
V1](https://github.com/vacp2p/specs/blob/master/specs/waku/waku-1.md) for privacy-preserving messaging.

> There MUST be a node that is capable of discovering peers and

Reviewer: This isn't accurate, for v1 of the app we still support Whisper. Instead, it should be MAY for Whisper and SHOULD for Waku. I also think the version in production is v1, but I might be wrong on that.

implements Waku V0 specification.

#### Node discovery and roles

There are four types of node roles:
1. Bootstrap nodes
2. Waku relayers
3. Mailservers (servers and clients)
4. Mobile nodes (Status Clients)

```js

> // TODO do waku relayers exist? Are they different from mobile nodes?

Reviewer: Should be essentially same as for Whisper. It'd be good if we kept Whisper in history here, so we have Whisper and Waku relayers. Then, at a later stage, we can deprecate Whisper completely.

cc @jakubgs re Waku relays in cluster

Protocol, v5](https://github.com/ethereum/devp2p/blob/master/rlpx.md) is used, which
allows for TCP-based communication between nodes.

On top of this we run RLPx-based subprotocols, the client
MAY use [Whisper v6](https://eips.ethereum.org/EIPS/eip-627), though the client 
SHOULD use [Waku V1](https://github.com/vacp2p/specs/blob/master/specs/waku/waku-1.md)

> for privacy-preserving messaging.

Reviewer: Perhaps add a sentence on why Waku should be used, like "Waku should be used because it preserves your bandwidth better". Additionally, we might want to say here that Whisper SHOULD NOT be used to indicate that it is a compatibility thing.

Author: ~~A point of clarification. Are you suggesting to change:~~

~~"MAY use Whisper" to "SHOULD NOT use Whisper"?~~

EDIT:

I'm presuming that is what you mean.

#### Node discovery and roles

There are five types of node roles:
1. `Bootstrap node`

> 1. `Whisper/Waku relayer`

Reviewer: Numbered list format off

Reviewer: Also where is the five from?

Author: Ah good catch, I had previously split Whisper and Waku relayers into separate node roles. I will change this.

I changed the numbered list so that all are `1.` because markdown parses `1.` as an ordered list. That means that additions to the list don't need to be manually re-sequenced.

Reviewer: Huh TIL. So this will render correctly for all (well, ours) kramdown generator?

Author: I had to check this, but it appears to already be something some contributors have done, and is working with the live specs.

See https://specs.status.im/spec/2#introduction

> 1 a whisper chat identity keypair
> 2 a set of cryptocurrency wallet keypairs

Which in the raw `.md` file looks like this :

https://github.com/status-im/specs/blob/56d5a2ff3949f3d2a68c70e1ae6780d2a080f16d/docs/stable/2-account.md#L51-L54

---

I just saw a better example:

https://specs.status.im/spec/2#private-11-messages

> 1 scanning a user generated QR code
> 2 discovery through the Status app
> 3 asynchronous X3DH key exchange
> 4 public key via public channel listening
>   . status-react/src/status_im/contact_code/core.cljs
> 5 contact codes
> 6 decentralized storage (not implemented)
> 7 whisper

In raw `.md` file looks like:

https://github.com/status-im/specs/blob/56d5a2ff3949f3d2a68c70e1ae6780d2a080f16d/docs/stable/2-account.md#L130-L139

This is a good example of how Markdown actually parses a list of `1.` mixed with other numbers. It is easier to just have all items as `1.` so no manual sequencing is needed.

1. `Mailserver` (servers and clients)
1. `Mobile node` (Status Clients)

> A standard Status client MUST implement `Whisper/Waku relayer` roles,

Reviewer: Hmm I realize this was in old code, but is this true? If you implement a mobile node, do you also need to implement a relay node? It is true we get forwarded envelopes on mobile (but don't relay further). What do we mean by mobile node in that case? Something seems off here

cc @kdeme @cammellos @decanus

Reviewer 2: Yeah, I guess it should rather be something like:
"A Status client MUST implement the `Mobile node` role and MAY implement `Whisper/Waku relayer` and `Mailserver` roles."

Reviewer 2: The whole role thing is not so clear though. I think it is because we are mixing Whisper/Waku & Status Client layers. And then also bootstrap..

Author: >If you implement a mobile node, do you also need to implement a relay node?

I puzzled over this question when I was drafting this PR.

>"A Status client MUST implement the Mobile node role and MAY implement Whisper/Waku relayer and Mailserver roles."

Are there circumstances when a Status node would be active and not be using p2p messaging?

Reviewer: Yeah, we need to clarify the roles a lot more. Since this issue exists in previous spec, could you open an issue for clarifying this @Samyoul? Then we can deal with it separately to get this merged. (I think there was some related roles issue but can't find it right now)

Author: https://github.com/status-im/specs/issues/123

Protocol, v5](https://github.com/ethereum/devp2p/blob/master/rlpx.md) is used, which
allows for TCP-based communication between nodes.

On top of this we run RLPx-based subprotocols, the client
MAY use [Whisper v6](https://eips.ethereum.org/EIPS/eip-627), though the client 
SHOULD use [Waku V1](https://github.com/vacp2p/specs/blob/master/specs/waku/waku-1.md)
for privacy-preserving messaging.

There MUST be a node that is capable of discovering peers and implements

> [Whisper v6](https://eips.ethereum.org/EIPS/eip-627),

Reviewer 2: The way it is written now sounds like you MUST implement all 3. While it is actually only 1.

Author: I intentionally wrote this so that there "MUST be **a** node" that implements all 3. I presumed that if there is a network that MAY implement Whisper, Waku 0 and Waku 1, then that network must have at least 1 node that implements all messaging protocols.

The other alternative is to make the language conditional and state that "There MUST be a node that ... implements all the messaging protocols that the network supports".

This would probably require a section defining the term "messaging protocol".

Reviewer 2: Right, I understand now, but this was not clear to me when I first read it.
Yes, I think it would be better to explain that part in another section where e.g. the bridging is also explained, or at least referenced from the vacp2p Waku spec.
Perhaps this part does not even belong in the client.md, as you do not have to do this to create a Status client.

Author: >Perhaps this part does not even belong in the client.md, as you do not have to do this to create a Status client.

Yes, I agree with this. Although `client.md` is home to a lot of details about different types of nodes required for a Status client to operate effectively.

[I think it is necessary for a Whisper / Waku bridge spec page](https://github.com/status-im/specs/pull/114#issuecomment-625887093), although the bridge spec would be fairly small, the details of the bridge spec would not neatly belong in any of the currently existing or proposed spec pages.

It would also allow for easy deprecation when required. I don't know if that would be overkill. What do you think?

Author: @kdeme what is your opinion?

Reviewer 2: Either works for me. Could also be just a subsection in the Whisper-usage.md. As it is bridging to support still Whisper. And can be removed easily then later on.

[Waku V0](https://github.com/vacp2p/specs/blob/master/specs/waku/waku-0.md) and
[Waku V1](https://github.com/vacp2p/specs/blob/master/specs/waku/waku-1.md) specifications.

#### Node discovery and roles

There are five types of node roles:
1. `Bootstrap node`

> 1. `Waku relayer`

Reviewer 2: Probably shouldn't create a new role for Waku relayer. Just call it Waku or Whisper Relayer (or just Relayer?).
As In theory then you also have to split Mailserver and Mobile node, they both still us the Waku / Whisper protocol

Author: In this section I was implementing a suggestion from @oskarth https://github.com/status-im/specs/pull/114#discussion_r421938443 .

While rewriting this section I had the idea to simply rename the `Whisper relayer` to `Message relayer` and add a definition of a`Message relayer`. I decided to explicitly keep both `Whisper relayer` and `Waku relayer` so that deprecation would be simpler.

Yes I may eventually split the `Mailserver`, but I haven't rewritten that section yet and I wanted to see how divergent Whisper vs Waku Mailservers were.

Could you explain why the `Mobile node` would need to be split?

Reviewer 2: I wouldn't split them, as the role is essentially the same, just a different protocol. As is the case for Mail server and mobile node. Yeah, perhaps Message relayer (Waku or Whisper) is better. Or just Waku/Whisper relayer.

Author: I like `Waku/Whipser relayer`, let's go with that.

1. `Whisper relayer`
1. `Mailserver` (servers and clients)
1. `Mobile node` (Status Clients)

> To implement a standard Status client you MUST implement `Waku relayer` and/or `Whisper relayer` roles,

Reviewer 3: ```suggestion
A standard Status client MUST implement `Waku relayer` and/or `Whisper relayer` roles,
```

I don't like the usage of you

Author: Neither do I.  There's quite a bit of the use of personal pronouns in these specs. I'll keep an eye out for them and replace them as I come across them.

> additionally the Status client MUST implement the `Mobile node` role.

Reviewer 2: I'm questioning the original sentence here. I think `Waku relayer` node is enough to implement a Status client. From the small explanation that is written about mobile nodes, I conclude this is the Waku light node. So, why would it not work with the full node, the light node is a subset.

Author: Should we have a concept of a `Mobile node`? Currently a `mobile node` is defined as:

>This is a Whisper and/or Waku node which connects to part of the respective Whisper
and/or Waku network(s). It MAY relay messages.

How does this diverge from a plain Whisper / Waku node ?

Reviewer 2: From the current description, or rather from the fact that "Status clients" is mentioned, it looks like the mobile node role is a node that not only speaks Waku/Whisper, but also the rest of the Status protocol to be able to send/receive chat messages (the layers on top of Whisper/Waku).

Compared to a Waku/Whisper relayer which doesn't need to know these layers to just pass the Waku/Whisper envelopes. I don't fully like the name mobile node, but ok, that is not the scope of this PR. So yeah, you can ignore my comment.

> The other node types are optional, but we RECOMMEND you implement a `Mailserver`

Reviewer 3: ```suggestion
The other node types are optional, but it is RECOMMEND to implement a `Mailserver`
```

Author: Probably should say 

The other node types are optional, but it is RECOMMENDED to implement a `Mailserver`

Protocol, v5](https://github.com/ethereum/devp2p/blob/master/rlpx.md) is used, which
allows for TCP-based communication between nodes.

On top of this we run RLPx-based subprotocols, the client
SHOULD NOT use [Whisper v6](https://eips.ethereum.org/EIPS/eip-627), the client 

> SHOULD use [Waku V1](https://github.com/vacp2p/specs/blob/master/specs/waku/waku-1.md)

Reviewer 2: nitpick: no capital v like in Whisper v6

Both peers discovery mechanisms use topics to provide peers with certain capabilities.
There is no point in returning peers that do not support a particular protocol.
Status nodes that want to be discovered MUST register to Discovery V5 and/or Rendezvous

> with the `whisper` and/or `waku` topic(s). Status nodes that are `Mailservers` and want to

Reviewer: Did we definitely change this to `waku` topic?

Author: I don't know. It is related to one of my outstanding issues. There is a reference in the specs to `whispermail`, but I haven't been able to find if it was changed to `wakumail` with Waku. 

In this section I presumed that Waku would have its `waku` topic set rather than still used the `whisper` topic.

Reviewer: Don't know. I'd grep the status-go code for this. If you can only find whispermail we didn't change it.

Reviewer: https://github.com/status-im/status-go/search?q=whispermail&unscoped_q=whispermail and https://github.com/status-im/status-go/search?q=wakumail&unscoped_q=wakumail seems like no

Author: That is a useful feature! Ok, I suppose that answers the "do we" question, the next question would be "should we?".

I'll remove the changes I've made here and I'll also break this into a separate issue.

Both peers discovery mechanisms use topics to provide peers with certain capabilities.
There is no point in returning peers that do not support a particular protocol.
Status nodes that want to be discovered MUST register to Discovery V5 and/or Rendezvous

> with the `waku` topic. Status nodes that are mail servers and want to

Reviewer: See above, MUST seems like too strong of a word and would also make this spec incompatible with v1 in wild, which is not what we want / are doing

Author: Ok. Thank you.

A `Mobile node` is a Whisper and/or Waku node which connects to part of the respective Whisper
and/or Waku network(s). A `Mobile node` MAY relay messages. See next section for more details on how
to use Whisper and/or Waku to communicate with other Status nodes.

### Transport privacy and Whisper / Waku usage

Once a Whisper and/or Waku node is up and running there are some specific settings required
to commmunicate with other Status nodes.

See [3/WHISPER-USAGE](https://specs.status.im/spec/3) and [9/WAKU-USAGE](https://specs.status.im/spec/9) for more details.

For providing offline inboxing, see the complementary [4/WHISPER-MAILSERVER](https://specs.status.im/spec/4) and [10/WAKU-MAILSERVER](https://specs.status.im/spec/10).

### Secure Transport

In order to provide confidentiality, integrity, authentication and forward
secrecy of messages we implement a secure transport on top of Whisper and Waku. This is
used in 1:1 chats and group chats, but not for public chats. See [5/SECURE-TRANSPORT](https://specs.status.im/spec/5) for more.

### Data Sync

[MVDS](https://specs.vac.dev/mvds.html) is used for 1:1 and group chats, however it is currently not in use for public chats.

> [Status payloads](#payloads-and-clients) are serialized and then wrapped inside an

Reviewer: This looks weird spacing-wise in GH at least

A `Mobile node` is a Whisper and/or Waku node which connects to part of the respective Whisper
and/or Waku network(s). A `Mobile node` MAY relay messages. See next section for more details on how
to use Whisper and/or Waku to communicate with other Status nodes.

### Transport privacy and Whisper / Waku usage

Once a Whisper and/or Waku node is up and running there are some specific settings required
to commmunicate with other Status nodes.

See [3/WHISPER-USAGE](https://specs.status.im/spec/3) and [9/WAKU-USAGE](https://specs.status.im/spec/9) for more details.

For providing offline inboxing, see the complementary [4/WHISPER-MAILSERVER](https://specs.status.im/spec/4) and [10/WAKU-MAILSERVER](https://specs.status.im/spec/10).

### Secure Transport

In order to provide confidentiality, integrity, authentication and forward
secrecy of messages we implement a secure transport on top of Whisper and Waku. This is
used in 1:1 chats and group chats, but not for public chats. See [5/SECURE-TRANSPORT](https://specs.status.im/spec/5) for more.

### Data Sync

[MVDS](https://specs.vac.dev/mvds.html) is used for 1:1 and group chats, however it is currently not in use for public chats.

> [Status payloads](#payloads-and-clients) are serialized and then wrapped inside a

Reviewer: spacing

#### Mobile nodes

This is a Waku node which connects to part of the Waku network. It MAY
relay messages. See next section for more details on how to use Waku to
communicate with other Status nodes.

### Transport privacy and Waku usage

Once a Waku node is up and running there are some specific settings required
to commmunicate with other Status nodes.

> See [3/WAKU-USAGE](https://specs.status.im/spec/3) for more details.

Reviewer: This spec doesnt exist

Author: My presumption is that this will reference https://github.com/status-im/specs/blob/master/docs/stable/3-whisper-usage.md, after the whole rewrite has completed.

Author: I'm going to revert this change, as per https://github.com/status-im/specs/pull/114#issuecomment-625894903

## Design Rationale

> ### P2P Overlay 1

Reviewer: Extra number?

Author: These things are my IDE. It forces unique anchors. I'll change it.

Status clients currently do not offer native support for file storage.

#### Why do you use Waku?

> Waku is one of the [three parts](http://gavwood.com/dappsweb3.html) of the

Reviewer 2: I guess this should remain Whisper.

Author: Yeah. I wasn't sure if we were treating Waku as the direct successor and replacement for Whisper, but it doesn't make sense in this sentence. I think it would be best to revert the change and remove this sentence when we phase Whisper our 100%.

For some further investigation, see
[here](https://github.com/status-im/swarms/blob/master/ideas/092-disc-v5-research.md).

#### I heard something about `Mailservers` being trusted somehow?

In order to use a `Mailserver`, a given node needs to connect to it directly, i.e. add the `Mailserver` as its peer and mark it as trusted. This means that the `Mailserver` is able to send direct p2p messages to the node instead of broadcasting them. Effectively, it knows the bloom filter of the topics the node is interested in, when it is online as well as many metadata like IP address.

> ### Data sync 1

Reviewer: Extra number?

large group contexts, which means bandwidth usage will be a lot higher than
reasonable. See [P2P Data Sync for
Mobile](https://vac.dev/p2p-data-sync-for-mobile) for more. This is an active
area of research.

> 

Reviewer: Could we add a CHANGELOG and bump version number somewhere here too?

## Appendix A: Security considerations

There are several security considerations to take into account when running Status. Chief among them are: scalability, DDoS-resistance and privacy. These also vary depending on what capabilities are used, such as `Mailserver`, light node, and so on.

### Scalability and UX

**Bandwidth usage:**

> In version 1 of Status, bandwidth usage is likely to be an issue. In version this is partially addressed with Waku usage, see [the theoretical scaling model](https://github.com/vacp2p/research/tree/dcc71f4779be832d3b5ece9c4e11f1f7ec24aac2/whisper_scalability).

Reviewer 2: I think you for got to add the actual version?

Author: Good catch.

### Scalability and UX

**Bandwidth usage:**

> In version 1 of Status, bandwidth usage is likely to be an issue. For more investigation into this, see [the theoretical scaling model](https://github.com/vacp2p/research/tree/dcc71f4779be832d3b5ece9c4e11f1f7ec24aac2/whisper_scalability).

Reviewer: I'd keep this as it is still relevant in version 1. Replace second sentence with "In version <X> this is partially addressed with Waku usage

Author: FYI, I also added a reference to this in the new `9-waku-usage.md` document, under the `topic-interest` references.

Use of gossip-based routing doesn't necessarily scale. It means each node can see a message multiple times, and having too many light nodes can cause propagation probability that is too low. See [Whisper vs PSS](https://our.status.im/whisper-pss-comparison/) for more and a possible Kademlia based alternative.

```js

> // TODO investigate update or Waku analogous ^^^

Reviewer 2: Nothing has changed in this aspect for Waku

By having a bloom filter where only the topics you are interested in are set, you reveal which messages you are interested in. This is a fundamental tradeoff between bandwidth usage and privacy, though the tradeoff space is likely suboptimal in terms of the [Anonymity](https://eprint.iacr.org/2017/954.pdf) [trilemma](https://petsymposium.org/2019/files/hotpets/slides/coordination-helps-anonymity-slides.pdf).

**`Mailserver` client privacy:**

> A `Mailserver` client has to trust a `Mailserver`, which means they can send direct traffic. This reveals what topics / bloom filter a node is interested in, along with its peerID (with IP).

Reviewer: Is there a reason Mailserver is in `` but `mailserver client` isn't?

Author: That makes more sense.
